### PR TITLE
#14665: add new moreh_clip_grad_norm and test in ttnn

### PR DESCRIPTION
**Pull request by DuongQLee**
### Ticket
Link to Github Issue: https://github.com/tenstorrent/tt-metal/issues/14665

### Problem description
moreh_clip_grad_norm is in tt-eager deprecated.

### What's changed
Currently, moreh_clip_grad_norm codes contain some bugs that can't be tested with normal test. I write a new test to ensure the output of the new version in ttnn produces similar output with the deprecated version. When the PR is approved, I will remove the deprecated moreh_clip_grad_norm in tt-eager and the newly added test.

### Checklist
- [x] Post commit CI passes: https://github.com/tenstorrent/tt-metal/actions/runs/11793801221
- [x] Blackhole Post commit (if applicable): NA
- [x] Model regression CI testing passes (if applicable): NA
- [x] Device performance regression CI testing passes (if applicable): NA
- [x] New/Existing tests provide coverage for changes: 18 tests passed

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step1/device/moreh_clip_grad_norm_step1_program_factory.cpp
Comment (o2buzzle) [inline ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step1/device/moreh_clip_grad_norm_step1_program_factory.cpp:141]: You can use `is_dram()` here instead of the ternary operation

File: ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step1/device/moreh_clip_grad_norm_step1_program_factory.cpp
Comment (o2buzzle) [inline ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step1/device/moreh_clip_grad_norm_step1_program_factory.cpp:155]: Similar to above

Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step1/device/moreh_clip_grad_norm_step1_program_factory.cpp]: Resolved in [367b454](https://github.com/tenstorrent/tt-metal/pull/14667/commits/367b454f49e71941dc71e9cf323bf305bcb799ba)

File: ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step1/device/moreh_clip_grad_norm_step1_program_factory.cpp
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step1/device/moreh_clip_grad_norm_step1_program_factory.cpp]: Resolved in [367b454](https://github.com/tenstorrent/tt-metal/pull/14667/commits/367b454f49e71941dc71e9cf323bf305bcb799ba)

File: ttnn/ttnn/operations/moreh.py
Comment (DuongQLee) [reply ttnn/ttnn/operations/moreh.py]: Resolved in [367b454](https://github.com/tenstorrent/tt-metal/pull/14667/commits/367b454f49e71941dc71e9cf323bf305bcb799ba)

**Review by o2buzzle — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/test_moreh_clip_grad_norm.py
Comment (hschoi4448) [inline tests/ttnn/unit_tests/operations/test_moreh_clip_grad_norm.py:224]: Could you delete this before getting it approved?

File: ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm.cpp
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm.cpp:83]:
```suggestion
        TT_FATAL(
```
Please, use TT_FATAL

File: ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm.cpp
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm.cpp:94]:
```suggestion
    Tensor scalar = creation::create_scalar(1.0f, inputs.at(0).get_dtype(), Layout::TILE, device);
```
I think "ttnn::operations::" is not necessary

File: ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step1/device/moreh_clip_grad_norm_step1_program_factory.cpp
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step1/device/moreh_clip_grad_norm_step1_program_factory.cpp:50]: please remove uncessary comment

**Comment by hschoi4448** [comment]
why not use compute_kernel_config in step1?

File: ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step1/device/kernels/moreh_clip_grad_norm_step1_kernel.cpp
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step1/device/kernels/moreh_clip_grad_norm_step1_kernel.cpp:61]: Please don't use ACQ() and REL().
it's deprecasted

File: ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step1/device/moreh_clip_grad_norm_step1_program_factory.cpp
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step1/device/moreh_clip_grad_norm_step1_program_factory.cpp:71]:
```suggestion
    TT_FATAL(core_group_2.ranges().empty());
    TT_FATAL(num_inputs_per_core_group_1 == 1);
    TT_FATAL(num_inputs_per_core_group_2 == 0);
```

File: ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step2/device/moreh_clip_grad_norm_step2_device_operation.cpp
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step2/device/moreh_clip_grad_norm_step2_device_operation.cpp:19]:
```suggestion
        check_tensor(tensor_args.total_norm, "moreh_clip_grad_norm_step2", "total_norm");
```

File: ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step2/device/moreh_clip_grad_norm_step2_program_factory.cpp
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step2/device/moreh_clip_grad_norm_step2_program_factory.cpp:66]:
```suggestion
    CreateCircularBuffer(
```

File: ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step2/device/moreh_clip_grad_norm_step2_program_factory.cpp
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step2/device/moreh_clip_grad_norm_step2_program_factory.cpp:91]:
```suggestion
    const auto reader_kernel_id = CreateReadKernel(program, reader_kernel_file, single_core);
    const auto writer_kernel_id = CreateWriteKernel(program, writer_kernel_file, single_core);
```

File: tests/ttnn/unit_tests/operations/test_moreh_clip_grad_norm.py
Comment (hschoi4448) [inline tests/ttnn/unit_tests/operations/test_moreh_clip_grad_norm.py:35]: Remove this and Please use `ttnn.to_torch`, `ttnn.from_torch`

File: tests/ttnn/unit_tests/operations/test_moreh_clip_grad_norm.py
Comment (hschoi4448) [inline tests/ttnn/unit_tests/operations/test_moreh_clip_grad_norm.py:166]:
```suggestion
        npu_total_norm = ttnn.moreh.clip_grad_norm(npu_inputs, max_norm, norm_type)
```
I think `operations.` is not neessary

File: tests/ttnn/unit_tests/operations/test_moreh_clip_grad_norm.py
Comment (DuongQLee) [reply tests/ttnn/unit_tests/operations/test_moreh_clip_grad_norm.py:224]: This test is included in the old test file so I copied it and change the deprecated moreh_clip_grad_norm operation to the new one to keep it consistent with the old version. Do you think I should delete it then?

**Review by hschoi4448 — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/test_moreh_clip_grad_norm.py
Comment (hschoi4448) [reply tests/ttnn/unit_tests/operations/test_moreh_clip_grad_norm.py:224]: I’m not sure about the details, but since GitHub keeps a history, there’s usually no reason to keep unused code.
However, since this PR isn’t specifically for refactoring, it should be fine to leave it as is without deleting.

File: ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm.cpp
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm.cpp]: Resolved in [06f7fe7](https://github.com/tenstorrent/tt-metal/pull/14667/commits/06f7fe78b8714d808e3442033a18a39ce88500c8)

File: ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm.cpp
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm.cpp]: Resolved in [06f7fe7](https://github.com/tenstorrent/tt-metal/pull/14667/commits/06f7fe78b8714d808e3442033a18a39ce88500c8)

File: ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step1/device/moreh_clip_grad_norm_step1_program_factory.cpp
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step1/device/moreh_clip_grad_norm_step1_program_factory.cpp]: Resolved in [06f7fe7](https://github.com/tenstorrent/tt-metal/pull/14667/commits/06f7fe78b8714d808e3442033a18a39ce88500c8)

File: ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step2/device/moreh_clip_grad_norm_step2_device_operation.cpp
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step2/device/moreh_clip_grad_norm_step2_device_operation.cpp]: Resolved in [06f7fe7](https://github.com/tenstorrent/tt-metal/pull/14667/commits/06f7fe78b8714d808e3442033a18a39ce88500c8)

File: ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step2/device/moreh_clip_grad_norm_step2_program_factory.cpp
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step2/device/moreh_clip_grad_norm_step2_program_factory.cpp]: Resolved in [06f7fe7](https://github.com/tenstorrent/tt-metal/pull/14667/commits/06f7fe78b8714d808e3442033a18a39ce88500c8)

File: ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step2/device/moreh_clip_grad_norm_step2_program_factory.cpp
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step2/device/moreh_clip_grad_norm_step2_program_factory.cpp]: Resolved in [06f7fe7](https://github.com/tenstorrent/tt-metal/pull/14667/commits/06f7fe78b8714d808e3442033a18a39ce88500c8)

File: tests/ttnn/unit_tests/operations/test_moreh_clip_grad_norm.py
Comment (DuongQLee) [reply tests/ttnn/unit_tests/operations/test_moreh_clip_grad_norm.py:95]: I will refactor the pytest when we fix the issue with the precision in later commit.

**Comment by DuongQLee** [comment]
I have removed deprecated moreh_clip_grad_norm code and test in https://github.com/tenstorrent/tt-metal/pull/14667/commits/174c0721cfa58970b469adbd630a33df475f6184

**Review by razorback3 — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm.hpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm.hpp:17]: Can we please replace this with enum class?
https://github.com/tenstorrent/tt-metal/blob/main/best_practices.md#14-avoid-bool-arguments-in-apis

File: ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm.cpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm.cpp:75]: Are you sure returning const& here is not an undefined behavior?

File: ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm.cpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm.cpp:94]: You store op result by const reference, but I think the returned value is a temporary object created on stack in those functions. I don’t think that a good idea.

**Review by ayerofieiev-tt — COMMENTED** [review]
A fairly large chunk of the test is marked as unsupported on BH or WH. 
Overall, the op seem to have many pieces and look understated. Specially the caching aspect of it. Please see if that can improved. Thank you

**Review by ayerofieiev-tt — APPROVED** [review]
Please review and address/resolve comments. Approving to not block once you make a pass.

File: ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm.cpp
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm.cpp]: resolved in commit [e1382c3](https://github.com/tenstorrent/tt-metal/pull/14667/commits/e1382c3076fc0b534597492750bfaa884a755fad). I have rerun the old test to make sure the output is still the same.

File: ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm.cpp
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm.cpp]: resolved in commit [e1382c3](https://github.com/tenstorrent/tt-metal/pull/14667/commits/e1382c3076fc0b534597492750bfaa884a755fad). I have rerun the old test to make sure the output is still the same.

**Pull request merged**
